src/backend/cdb/cdbmutate.c

```diff
@@ -936,17 +1057,11 @@ apply_motion_mutator(Node *node, ApplyMotionState *context)
 			break;
 
 		case MOVEMENT_EXPLICIT:
-
 			/*
-			 * add an ExplicitRedistribute motion node only if child plan
-			 * nodes have a motion node
+			 * Were there any motions above the scan?
 			 */
-			if (context->containMotionNodes)
+			if (context->mt.needExplicitMotion)
```

Comment:
Why do we need the mt.needExplicitMotion variable? Maybe replace it with a condition?
```suggestion
			if (context->mt.nMotionsAbove > 0)
```

Reply:
https://github.com/arenadata/gpdb/blob/a270ca14a014dbdf47b6432a964d8c2b7489a180/src/backend/cdb/cdbmutate.c#L932

This is a recursive function, so we first descend down the tree and count the motions. When we go out of recursion, the motion count is decremented back to 0. That means we need to save the state of the counter specifically when we encounter the scan, because it will be back to 0 when we encounter the explicit locus under ModifyTable.